### PR TITLE
ENH: Improve typing for pivot and pivot_table

### PR DESCRIPTION
**Pull request by bashtage**
- [ ] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

**Comment by bashtage** [comment]
@twoertwein any ideas about the pyright failure?

**Comment by twoertwein** [comment]
> @twoertwein any ideas about the pyright failure?

Sory for the late reply, I assume you fixed the error as pyright seems to pass now (but a pytest error).

File: pandas-stubs/core/reshape/concat.pyi [hunk 1]
```diff
@@ -12,15 +15,18 @@ from pandas import (
 
 from pandas._typing import HashableT
 
+_HashableT2 = TypeVar("_HashableT2", bound=Hashable)
+_HashableT3 = TypeVar("_HashableT3", bound=Hashable)
```
Comment (Dr-Irv) [inline pandas-stubs/core/reshape/concat.pyi:19]: Now that I merged in #377 you should use those

File: pandas-stubs/core/reshape/concat.pyi [hunk 2]
```diff
@@ -12,15 +15,18 @@ from pandas import (
...
+
 @overload
 def concat(
     objs: Iterable[DataFrame] | Mapping[HashableT, DataFrame],
     axis: Literal[0, "index"] = ...,
     join: str = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/reshape/concat.pyi:25]: `join` is `Literal["inner", "outer"]`

File: pandas-stubs/core/reshape/concat.pyi [hunk 3]
```diff
@@ -12,15 +15,18 @@ from pandas import (
...
     ignore_index: bool = ...,
-    keys=...,
-    levels=...,
-    names=...,
+    keys: Sequence[_HashableT2] = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/reshape/concat.pyi:27]: I think a `str` would match, so should be `list[HashableT2] | tuple[HashableT2, ...]`

File: pandas-stubs/core/reshape/concat.pyi [hunk 4]
```diff
@@ -12,15 +15,18 @@ from pandas import (
...
+    levels: list[Sequence] = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/reshape/concat.pyi:28]: `list[list[HashableT2] | tuple[HashableT2, ...]]` might be better

File: pandas-stubs/core/reshape/pivot.pyi [hunk 5]
```diff
@@ -1,36 +1,130 @@
...
 from pandas._typing import (
-    IndexLabel,
+    Label,
     Scalar,
+    npt,
 )
 
+_HashableT1 = TypeVar("_HashableT1", bound=Hashable)
+_HashableT2 = TypeVar("_HashableT2", bound=Hashable)
+_HashableT3 = TypeVar("_HashableT3", bound=Hashable)
```
Comment (Dr-Irv) [inline pandas-stubs/core/reshape/pivot.pyi:29]: Now you can use the ones in `_typing.pyi`

File: pandas-stubs/core/reshape/pivot.pyi [hunk 6]
```diff
@@ -1,36 +1,130 @@
...
+    Callable[[Series], datetime.date],
+    Callable[[Series], datetime.datetime],
+    Callable[[Series], datetime.timedelta],
+    Callable[[Series], bool],
+    Callable[[Series], int],
+    Callable[[Series], float],
+    Callable[[Series], complex],
+    Callable[[Series], pd.Timestamp],
+    Callable[[Series], pd.Timedelta],
+]
```
Comment (Dr-Irv) [inline pandas-stubs/core/reshape/pivot.pyi:42]: I'm pretty sure you can use `Callable[[Series], ScalarT]` here as `ScalarT` is a bound `TypeVar` of `Scalar`

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
On `concat`, I commented on the first overload, need to make sure to change all the overloads.

Not sure if the tests catch all cases, but that's really hard to do.

Might want to change some of the `np.sum` uses of `aggfunc` in the tests to things like `"sum"` , `"mean"`, etc.

File: pandas-stubs/core/reshape/concat.pyi (hunk 1 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/concat.pyi]: Fixed everywhere.

File: pandas-stubs/core/reshape/concat.pyi (hunk 2 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/concat.pyi]: Fixed.

File: pandas-stubs/core/reshape/concat.pyi (hunk 1 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/concat.pyi]: Fixed.

File: pandas-stubs/core/reshape/concat.pyi (hunk 2 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/concat.pyi]: Fixed.

File: pandas-stubs/core/reshape/concat.pyi (hunk 4 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/concat.pyi]: Went with `Sequence[list[HashableT2] | tuple[HashableT2, ...]` due to invariance.

File: pandas-stubs/core/reshape/pivot.pyi (hunk 5 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/pivot.pyi]: Fixed.

File: pandas-stubs/core/reshape/pivot.pyi (hunk 6 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/pivot.pyi]: Changed this.

File: pandas-stubs/core/reshape/concat.pyi (hunk 3 above)
Comment (bashtage) [reply pandas-stubs/core/reshape/concat.pyi]: Fixed.

File: tests/test_pandas.py [hunk 7]
```diff
@@ -90,20 +105,121 @@ def test_types_concat() -> None:
...
+    )
+
+    # @overload
+    # def concat(
+    #         objs: Iterable[DataFrame] | Mapping[HashableT, DataFrame],
+    #         keys=...,
+    #         levels=...,
+    #         names=...,
+
+    # ) -> DataFrame: ...
```
Comment (Dr-Irv) [inline tests/test_pandas.py:222]: should delete the above comments

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
just some extra comments that you should remove, otherwise LGTM

File: tests/test_pandas.py (hunk 7 above)
Comment (bashtage) [reply tests/test_pandas.py]: Removed.

**Comment by bashtage** [comment]
@Dr-Irv it is green now.

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

**Pull request merged**